crates/subspace-farmer/src/utils/farmer_piece_cache.rs

```diff
@@ -40,19 +43,24 @@ impl FarmerPieceCache {
             }
         }
 
-        Self { store, heap }
+        Self {
+            store,
+            heap: Arc::new(Mutex::new(heap)),
+        }
     }
 }
 
 impl PieceCache for FarmerPieceCache {
+    type KeysIterator = impl IntoIterator<Item = Key>;
+
     fn should_cache(&self, key: &Key) -> bool {
-        self.heap.should_include_key(key)
+        self.heap.lock().should_include_key(key)
     }
 
     fn add_piece(&mut self, key: Key, piece: Piece) {
```

Comment:
Hmm... This looks a bit weird. I believe right now none of the `FarmerPieceCache` fields require `&mut`. I believe this trait can be implemented for `&FarmerPieceCache`?

Reply:
Trait requires `&mut`, that is why we have it here as well. But the data structure itself due to use cases requires to be clonable, hence `Arc<Mutex<T>>`

Reply:
Yeah, trait does, but `&mut &T` is the same as `&T`. I'm saying that this trait can be implemented on `&FarmerPieceCache` in the end.

Reply:
Now we could, yes, but we typically implement traits on `T`, not `&T` for convenience.